SAGESharp/SLB/IO/BinaryReader.cs

```diff
@@ -0,0 +1,183 @@
+            => new StreamBinaryReader(stream);
+    }
+
+    internal sealed class StreamBinaryReader : IBinaryReader
+    {
+        private readonly Stream stream;
+
+        public StreamBinaryReader(Stream stream)
+            => this.stream = stream ?? throw new ArgumentNullException();
+
+        public long Position
+        {
+            get => stream.Position;
+            set => stream.Position = value;
+        }
+
+        public byte ReadByte() => ReadBytes(1)[0];
+
+        public byte[] ReadBytes(int count)
+        {
+            if (count == 0)
+            {
+                return Array.Empty<byte>();
+            }
+
+            var buffer = new byte[count];
+            if (stream.Read(buffer, 0, count) == count)
+            {
+                return buffer;
+            }
+            else
+            {
+                throw new EndOfStreamException();
+            }
+        }
+
+        public short ReadInt16()
+            => BitConverter.ToInt16(ReadBytes(2), 0);
+
+        public ushort ReadUInt16()
+            => BitConverter.ToUInt16(ReadBytes(2), 0);
+
+        public int ReadInt32()
+            => BitConverter.ToInt32(ReadBytes(4), 0);
+
+        public uint ReadUInt32()
+            => BitConverter.ToUInt32(ReadBytes(4), 0);
+
+        public float ReadFloat()
+            => BitConverter.ToSingle(ReadBytes(4), 0);
+
+        public double ReadDouble()
+            => BitConverter.ToDouble(ReadBytes(8), 0);
+    }
+
+    internal static class IBinaryReaderExtensions
+    {
+        /// <summary>
+        /// Executes <paramref name="action"/> while temporarily moving the reader to <paramref name= "position" />.
+        /// </summary>
+        /// 
+        /// <param name="reader">The reader that w</param>
+        /// <param name="position">The position where the reader will be moved temporarily.</param>
+        /// <param name="action">The action to execute.</param>
+        public static void OnPositionDo(this IBinaryReader reader, uint position, Action action)
+        {
+            var originalPosition = reader.Position;
+            reader.Position = position;
+
+            action();
+            reader.Position = originalPosition;
+        }
+
+        /// <summary>
+        /// Returns the result of <paramref name="function"/> while temporarily moving the reader to <paramref name= "position" />.
+        /// </summary>
+        /// 
+        /// <typeparam name="TResult">The type that will be returned.</typeparam>
+        /// 
+        /// <param name="reader">The reader that will be changed temproarily.</param>
+        /// <param name="position">The position where the reader will be moved temporarily.</param>
+        /// <param name="function">The action to execute.</param>
+        /// 
+        /// <returns>The result of <paramref name="function"/>.</returns>
+        public static TResult OnPositionDo<TResult>(this IBinaryReader reader, uint position, Func<TResult> function)
```

Comment:
I think `OnPositionDo` could possibly be changed to a name more descriptive of what the function does, such as `DoAtPosition`. Starting with `On` makes it sound like an event handler of some sort.

Reply:
Done